File: apps/device_controller/device_controller.lua

local bus_client = require('lib.bus.client')
local bus_server = require('lib.bus.server')

local comm = bus_client:new({
    id='device_controller',
    filter='home_stack'
})
local device_interface = bus_server:new({
    id='device_controller',
    com_port=5564,
    set_port=5565,
    res_port=5566
})

local running = true
local log = function(...) print('<:> Device Controller <:>', ...) end

comm:setup()
device_interface:setup()

local treat_comm_msg = function(msg)
    log('Message: from,type,data', msg.from, msg.type, msg.data)
    if msg.type == 'get' then 
        local my_response = 'my response to '..msg.data..' is buga.'
        comm:send(my_response, 'response')
    end
end

local treat_device_interface_msg = function(msg)
    log('Device Interface Message: ', msg.from, msg.type, msg.data)
    ui:distribute({from='device_controller', type=msg.type, data=msg.data})
end

local life_loop = function()
    local msg
    msg = comm:check_income()
    if msg then treat_comm_msg(msg); msg = nil end
    msg = device_interface:getMessage('noblock')
    if msg then treat_device_interface_msg(msg) end
end

log('Running ...')
while running do life_loop() end
log('Stopped.')

File: apps/device_controller/device_controller.py

import logging
try:
    from client import BusClient
except ImportError:
    print('WARNING: Please set PYTHONPATH correcly. See readme for more info.')

from zwave_controller import ZWaveController

running = True
logging.basicConfig(level=logging.DEBUG)

def lifeLoop():
    msg = home_stack.check_income()
    if msg:
        onHomeStackMessage(msg[0]);

def onHomeStackMessage(msg):
    logging.info('Message: from,type,data', extra=msg)
    if msg['type'] == 'get':
        if msg['data'] == 'devicelist':
            devicelist = zwave.getDeviceList()
            home_stack.send(devicelist, 'response')
    elif msg['type'] == 'send':
        id_device = msg['data']['id']
        mtype = msg['data']['type']
        state = msg['data']['state']
        if mtype == 'setstate':
            zwave.setDeviceState(id_device, state)
            print('New state of ' + str(id_device) + ' is ' + state)

def onDeviceUpdate(device):
    print('will send notifycation!')
    msg = {'type':'update','data':device,'id_device':device['id_device']}
    home_stack.send(msg, 'send')

home_stack = BusClient('device_controller', 'home_stack')
home_stack.setup()

zwave = ZWaveController()
zwave.setup(onDeviceUpdate)

logging.info('Running ...')
while running:
    lifeLoop()
logging.info('Stopped.')

